File: app/src/main/java/ru/skinallergic/checkskin/components/fooddiary/components/AddFoodFragment.java
package ru.skinallergic.checkskin.components.fooddiary.components;

import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import java.util.Date;

import ru.skinallergic.checkskin.R;
import ru.skinallergic.checkskin.databinding.FragmentAddFoodBinding;
import ru.skinallergic.checkskin.view_models.DateViewModel;
import ru.skinallergic.checkskin.Loger;
import ru.skinallergic.checkskin.components.fooddiary.viewModels.AddViewModel;
import ru.skinallergic.checkskin.components.fooddiary.viewModels.FoodDiaryViewModel;

public class AddFoodFragment extends Fragment {
    private String[] array;
    private AddViewModel viewModel;
    private FragmentAddFoodBinding binding;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        array=getResources().getStringArray(R.array.array_teme_of_meals);
        viewModel = new ViewModelProvider(requireActivity()).get(AddViewModel.class);
        FoodDiaryViewModel foodDiaryViewModel=new ViewModelProvider(requireActivity()).get(FoodDiaryViewModel.class);
        DateViewModel dateViewModel=new ViewModelProvider(requireActivity()).get(DateViewModel.class);
        binding=FragmentAddFoodBinding.inflate(inflater);
        binding.setFragment(this);
        binding.setViewModel(viewModel);
        View view=binding.getRoot();
        dateViewModel.dateLive.observe(getViewLifecycleOwner(),(Date d)-> {
            String date=dateViewModel.getDate(d);
            binding.date.setText(date);
            viewModel.date=date;
        });

        Spinner spinner = binding.spinner;
        ArrayAdapter<String> adapter = new ArrayAdapter<String>(getActivity(), R.layout.contact_spinner_row_nothing_selected) {

            @Override
            public View getView(int position, View convertView, ViewGroup parent) {

                View v = super.getView(position, convertView, parent);
                if (position == getCount()) {
                    ((TextView)v.findViewById(android.R.id.text1)).setText("");
                    ((TextView)v.findViewById(android.R.id.text1)).setHint(getItem(getCount())); //"Hint to be displayed"
                }

                return v;
            }

            @Override
            public int getCount() {
                return super.getCount()-1; // you dont display last item. It is used as hint.
            }
        };

        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        adapter.addAll(array);

        spinner.setAdapter(adapter);
        try {
            int position=getArguments().getInt("position");
            spinner.setSelection(position);
        }
        catch (NullPointerException exception){
            spinner.setSelection(adapter.getCount()); //set the hint the default selection so it appears on launch.
        }

        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                if (position==4){
                    viewModel.timeOfMeal="";
                } else {
                    viewModel.timeOfMeal=array[position];
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        return view;
    }
    public void backStack(View view){
        Navigation.findNavController(view).popBackStack();
    }
    public void editName(CharSequence s,int start, int count, int after){
        viewModel.name=s.toString();
    }
    public void editWeight(CharSequence s,int start, int count, int after){
        viewModel.weight=s.toString();
    }
    public void addPosition(View view){
        if (viewModel.validatePos()){
            viewModel.addposition()
            .subscribe((Boolean isSaved)->{
                //Clean edit Texts
                binding.editName.setText("");
                binding.editWeight.setText("");
                Toast.makeText(getContext(),"Позиция добавлена", Toast.LENGTH_LONG).show();
            });

        } else {
            Loger.log("Заполните все поля");
            Toast.makeText(getContext(),"Заполните все поля", Toast.LENGTH_LONG).show();
        }
    }
}

File: app/src/main/java/ru/skinallergic/checkskin/components/fooddiary/components/AddFoodFragment.kt
package ru.skinallergic.checkskin.components.fooddiary.components

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.widget.doAfterTextChanged
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView
import ru.skinallergic.checkskin.Loger
import ru.skinallergic.checkskin.components.fooddiary.adapters.DiffUtilFromMySelf
import ru.skinallergic.checkskin.components.fooddiary.adapters.RecyclerProductAdapter
import ru.skinallergic.checkskin.components.fooddiary.data.ProductEntity
import ru.skinallergic.checkskin.components.fooddiary.viewModels.AddingFoodViewModel
import ru.skinallergic.checkskin.components.fooddiary.viewModels.MealViewModel
import ru.skinallergic.checkskin.components.fooddiary.viewModels.add
import ru.skinallergic.checkskin.components.fooddiary.viewModels.delete
import ru.skinallergic.checkskin.databinding.FragmentAddFood2Binding
import java.util.*

class AddFoodFragment : BaseFoodFragment(){
    val eatingList=listOf("Выберите прием пищи", "Завтрак", "Обед", "Перекус", "Ужин")
    lateinit var backStack: ImageButton
    lateinit var saveButton: ImageButton
    lateinit var spinner: Spinner
    lateinit var addButton: Button
    lateinit var binding: FragmentAddFood2Binding
    lateinit var recyclerView: RecyclerView
    lateinit var adapter :RecyclerProductAdapter
    lateinit var thisView: View
    val viewModel by lazy { ViewModelProvider(this, viewModelFactory).get(AddingFoodViewModel::class.java) }
    val mealViewModel by lazy { ViewModelProvider(this,viewModelFactory).get(MealViewModel::class.java) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        adapter= RecyclerProductAdapter()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentAddFood2Binding.inflate(inflater)
        thisView= binding.root
        binding.apply {
            backStack=backBtn
            saveButton=okBtn
            spinner=eatTime
            addButton=buttonAdd
            fragment=this@AddFoodFragment
            recyclerView=recycler
            baseViewModel=mealViewModel
        }
        recyclerView.adapter=adapter

        spinner.setAdapter(eatingList)
        subscribeDate()
        return thisView
    }

    override fun onStart() {
        super.onStart()

        spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener{
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                if (position!=0){
                    println(eatingList[position])
                    when(position){
                        1->mealViewModel.meal=0 //перекодирование для сервера
                        2->mealViewModel.meal=1
                        3->mealViewModel.meal=3
                        4->mealViewModel.meal=2
                    }
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>?) { }
        }

        addButton.setOnClickListener {
            if (viewModel.conditionOfAdding(getData())){
                add(ProductEntity())
            } else {toastyManager.toastyyyy("Пожалуйста заполните предыдущую позицию")}
        }

        adapter.apply{
            setDeletingFunction { position-> delete(position) }
            bind {holder,entity->
                holder.name.text= entity.name
                holder.weight.text= entity.weight
                holder.delete.setOnClickListener { deletingFun(entity) }
                holder.name.doAfterTextChanged { entity.name=it.toString(); deleteButtonVisibility(holder, entity) }
                holder.weight.doAfterTextChanged { entity.weight=it.toString() }
            }
        }

        add(ProductEntity())
        mealViewModel.productList.observe(viewLifecycleOwner, {newData->
            val diffUtil = DiffUtilFromMySelf<ProductEntity>(adapter.getData(), newData)
            diffUtil.calculate(adapter)
        })

        saveButton.setOnClickListener {
            //viewModel.save(adapter.getData())
            Loger.log("setOnClickListener")
            save()
        }

        backStack.setOnClickListener {
            Loger.log("mealViewModel.quitSaveCondition() ${mealViewModel.quitSaveCondition()}")
            quitSaveLogic({
                mealViewModel.quitSaveCondition() }, {
                popBack(it) }, {
                //viewModel.backSave(adapter.getData())
                save(mealViewModel.isBackSaved)
                })
        }

        /*viewModel.isBackSaved.observe(viewLifecycleOwner,{ isSaved->
            if (isSaved){popBack(thisView)}
        })*/

        mealViewModel.isBackSaved.observe(viewLifecycleOwner,{ isSaved->
            if (isSaved){
                mealViewModel.isBackSaved.value=false
                popBack(thisView) // view ??????????????????????????????????
            }
        })
    }

    fun deleteButtonVisibility(holder: RecyclerProductAdapter.Item, entity: ProductEntity){
        if (viewModel.conditionOfDelete(entity)){
            holder.deleteBtnVisible.set(true)
        } else holder.deleteBtnVisible.set(false)
    }

    fun quitSaveCondition(): Boolean {
        return false
    }

    fun add(position: ProductEntity){
        mealViewModel.productList.add(position)
    }

    fun delete(position: ProductEntity){
        mealViewModel.productList.delete(position)
    }

    fun getData():List<ProductEntity>{
        return adapter.getData()
    }

    fun save(backSaved: MutableLiveData<Boolean>?=null){
        Loger.log("save")
        val list: List<ProductEntity> =adapter.getData()
        mealViewModel.productList.value
        if (mealViewModel.saveCondition()){
            mealViewModel.addMealsAndConvert(
                    dateViewModel.dateUnix,
                    mealViewModel.meal!!,
                    list,
                    backSaved
            )
        }
    }
}